mets_adapter/mets_adapter/src/main/scala/uk/ac/wellcome/mets_adapter/services/MetsStore.scala

```diff
@@ -0,0 +1,27 @@
+package uk.ac.wellcome.mets_adapter.services
+
+import uk.ac.wellcome.mets_adapter.models._
+import uk.ac.wellcome.storage.store.VersionedStore
+import uk.ac.wellcome.storage.{Identified, StorageError, NoVersionExistsError}
+
+/** Stores METS data only if a newer version and the METS XML location is pointing
+ *  to a different file, returning the stored data if it has been updated.
+ */
+class MetsStore(store: VersionedStore[String, Int, MetsData]) {
+
+  def storeMetsData(bagId: String, data: MetsData): Either[Throwable, Option[MetsData]] =
+    shouldUpdate(bagId, data)
+      .right.flatMap {
+        case true => store.putLatest(bagId)(data).right.map(_ => Some(data))
+        case false => Right(None)
+      }
+      .left.map(_.e)
```

Comment:
I have thought of an obscure failure which could occur. If we successfully store the `MetsData`, but then encounter an error when publishing the message to SNS, the message will go back on the queue. However, the next time round when we call `shouldUpdate` it will look like we have already dealt with this data, so we don't bother storing it and publishing the message (leading to the data being lost and never sent to the transformer).

The only ways I can think of a way to getting around this apart are:
* We still publish the message when we receive another one containing the same version
* Introduce some complicated transaction logic so the dynamo table is rolled back if publishing the message to SNS fails (pretty sure the storage libs don't support such a thing out of the box)

:thinking: any thoughts @alicefuzier ?

Reply:
Another option:

* First check whether the data is stored within dynamo, then publish to SNS, then actually store the data in dynamo: this could lead to the same message being published to SNS multiiple times storing in dynamo fails, but I assume this is preferable to having data never make it into the pipeline.

Reply:
I have implemented the bottom option above :point_up: